src/table.rs

```diff
@@ -15,6 +15,9 @@ pub struct Table {
     pub(crate) decor: Decor,
     // whether to hide an empty table
     pub(crate) implicit: bool,
+    // used for putting tables back in their original order when serialising.
+    // Will be None when the Table wasn't parsed from a file.
+    pub(crate) position: Option<usize>,
```

Comment:
The `usize` can actually be anything that implements PartialOrd, if you have a preference.

Reply:
`usize` seems fine to me.

I'm thinking of weird ways of abusing this field. What happens if we have two documents and we `mem::swap` tables from different documents with different indices? We could end up with two table with the same position.

Reply:
of even swappin a new table with an existing table in the document would change the order, right?

Reply:
If we end up with two tables in the same position then we will print them one after the other (a previous implementation in this pr would silently drop them)

If you mem::swap two tables from the same doc like the implicit tables a and z in here:
```
[a.b]
[z.x]
```
Then Display will format it like this because it cares most about the order of entries in the top level Document
```
[a.x]
[z.b]
```
in_original_order will print:
```
[z.b]
[a.x]
```
because it skips the implicit tables and puts b and x in their original order.

This is a bit of a contrived example though.

The more interesting case is if you parse two Documents and then try to append the top level keys from the second doc onto the first. In the current implementation, in_original_order will interleave the tables from each doc, so I would recommend using Display::to_string() instead. I'm not sure it's well defined which order you want though, because both functions will hoist non-Table values to the top. If you don't want interleaving then you could make position be a tuple of (document_id, position), where document_id is an integer that increments whenever you start parsing a document. I'm not sure it's worth the added complexity though, because I don't think anyone will ever want to do this and also care about honouring split tables.

I suppose that any "this is an edge case" argument I make is weakened by the fact that this entire pr is about fixing a 2% edge case, so I will take your lead on this.

Reply:
If we end up with two tables in the same position then we will print them one after the other (a previous implementation in this pr would silently drop them)

If you mem::swap two tables from the same doc like the implicit tables a and z in here:
```
[a.b]
[z.x]
```
Then Display will format it like this because it cares most about the order of entries in the top level Document
```
[a.x]
[z.b]
```
in_original_order will print:
```
[z.b]
[a.x]
```
because it skips the implicit tables and puts b and x in their original order.

This is a bit of a contrived example though.

The more interesting case is if you parse two Documents and then try to append the top level keys from the second doc onto the first. In the current implementation, in_original_order will interleave the tables from each doc, so I would recommend using Display::to_string() instead. I'm not sure it's well defined which order you want though, because both functions will hoist non-Table values to the top. If you don't want interleaving then you could make position be a tuple of (document_id, position), where document_id is an integer that increments whenever you start parsing a document. I'm not sure it's worth the added complexity though, because I don't think anyone will ever want to do this and also care about honouring split tables.

I suppose that any "this is an edge case" argument I make is weakened by the fact that this entire pr is about fixing a 2% edge case, so I will take your lead on this.